### PR TITLE
Expr: Support for protobuf for MirScalarExpr.

### LINKED ISSUE #11745: Add protobuf support for `mz_expr::scalar::func::*`
Subtasks:
  - [x] #11746
  - [x] #11747
  - [x] #11748
  - [x] #11749

### LINKED ISSUE #11747: `mz_expr::scalar::func::UnaryFunc` — 213 variants

### PR DESCRIPTION
**Pull request by wangandi**
Relates to #11740.
Closes #11745.
Closes #11747.

TODO before closing #11740: 
* Have full coverage for `MirScalarExpr` proptesting.

### Tips for reviewer

`From` and `TryInto` for `UnaryFunc` variants that depend on `MirScalarExpr` have now been filled in.

There is a fair amount of refactoring to get rid of recursive imports:
* The protobuf versions of all `expr::scalar::*Func` types have been moved to src/scalar.proto.
* The protobuf versions of all `expr::relation::*Func` types have been moved to src/relation.proto.

### Testing

- [ ] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

No user-facing behavior changes.

### CODE REVIEW

File: src/expr/src/scalar/mod.rs [hunk 1]
```diff
@@ -77,28 +78,147 @@ pub enum MirScalarExpr {
 }
 
 impl Arbitrary for MirScalarExpr {
-    // TODO: This is just a stub. #11740
     type Parameters = ();
-    type Strategy = Just<MirScalarExpr>;
+    type Strategy = BoxedStrategy<MirScalarExpr>;
```
Comment (lluki) [inline src/expr/src/scalar/mod.rs:82]: Have you tried if adding `Union<>` fixes the stackoverflow issue? (It's fine to me to merge it without tho)

File: src/expr/src/scalar/mod.rs [hunk 2]
```diff
@@ -77,28 +78,147 @@ pub enum MirScalarExpr {
...
 
     fn arbitrary_with(_: Self::Parameters) -> Self::Strategy {
-        prop_oneof![Just(MirScalarExpr::Column(0))]
```
Comment (lluki) [inline src/expr/src/scalar/mod.rs:85]: Can you add add a TODO comment here?

**Review by lluki — APPROVED** [review]

File: src/expr/src/scalar/mod.rs (hunk 1 above)
Comment (wangandi) [reply src/expr/src/scalar/mod.rs:82]: I managed to fix the stack overflow issue by manually implementing `Arbitrary` for `AggregateFunc`.

File: src/expr/src/scalar/mod.rs [hunk 3]
```diff
@@ -77,28 +78,147 @@ pub enum MirScalarExpr {
...
+impl TryFrom<proto_literal::ProtoLiteralData> for Result<Row, EvalError> {
+    type Error = TryFromProtoError;
+
+    fn try_from(value: proto_literal::ProtoLiteralData) -> Result<Self, Self::Error> {
+        use proto_literal::proto_literal_data::Result::*;
+        match value.result {
+            Some(Ok(row)) => Result::Ok(Result::Ok(
+                (&row)
+                    .try_into()
+                    .map_err(TryFromProtoError::RowConversionError)?,
```
[the remark this answers is not in the document]
Comment (wangandi) [reply src/expr/src/scalar/mod.rs:205]: Created issue #12125 for this.

File: src/expr/src/scalar/mod.rs (hunk 2 above)
Comment (wangandi) [reply src/expr/src/scalar/mod.rs:85]: I filled in the entire implementation for MirScalarExpr, so that is no longer necessary.

**Comment by wangandi** [comment]
Re-requesting review because I managed to fix the stack overflow and fill out all of the Arbitrary implementation for MirScalarExpr except for MirScalarExpr::literal_ok. See the two new commits.

**Review by lluki — APPROVED** [review]

**Pull request merged**